articles/102_dependency_groups.md

```diff
@@ -0,0 +1,168 @@
+---
+layout: default
+title: Dependency groups
+permalink: articles/dependency_groups.html
+abstract:
+  This article describes the concept of dependency groups, their use cases, how they can described in the package manifest, and how they can be mapped into binary packages.
+author: '[Dirk Thomas](https://github.com/dirk-thomas)'
+published: true
+categories: Overview
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Preface
+
+In the ROS ecosystem a ROS package provides information about its dependencies in the manifest.
+The latest format is specified in the [REP 140](http://www.ros.org/reps/rep-0140.html).
+It supports different kinds of dependencies (`build`, `exec`, `test`, `doc`, etc.) but each dependency is mandatory and needs to satisfied on its own.
+
+This article describes the concept of "dependency groups" which could be used for different purposes.
+In terms of terminology a dependency group is a set of dependencies identified by the name of the group.
+For such a group a different semantic can then be applied (instead of just satisfying each dependency on its own).
+
+## Use Cases
+
+The following subsections will describe how dependencies could be used in two different use cases.
+
+### (1) Need at least one "Provider"
+
+A package might want to declare that at least one dependency from a group needs to be present (called `satisfy-at-least-one`).
+Compared to "normal" dependencies it is not required that all dependencies of that group need to be present to satisfy the dependency group.
+
+An example for this is the `rmw_implementation` package.
+Currently it depends on all RMW implementation packages (e.g. `rmw_fastrtps_cpp`, `rmw_connext_cpp`, etc.) in order to ensure that it may only be built after all RMW implementation packages have been built.
+The goal would be that at least one RMW implementation needs to be present in order to build or use the `rmw_implementation` package.
+It would be acceptable if more than one or even all RMW implementations are available but that is not required.
```

Comment:
does this imply the responsibility of the user that the semantics of the build does not change when a second item of the group is found afterwards? I am imagining the scenario where we find only one RMW, start building with the "shortcut" version, then figuring out a second RMW impl is available, but `rmw_implementation` is already setup wo/ dlopen support. Does this make sense?

Reply:
From source all rmw impl. in the workspace will be build before the `rmw_implementation` package. If the workspace only contains one and later you build another workspace on top with another rmw implementation that wouldn't change `rmw_implementation`.

When building Debian packages at least one needs to be available. But more could be available.

Reply:
After getting a bit more context offline, I got a few more insights which resolves my question here. The key point was that still all packages are traversed by ament before the first package is built.